src/renderer/tr_bsp.cpp

```diff
@@ -230,16 +248,53 @@ static	void R_LoadLightmaps( lump_t *l, const char *psMapName ) {
 				image[j*4+3] = 255;
 			}
 		}
-		tr.lightmaps[i] = R_CreateImage( va("*%s/lightmap%d",sMapName,i), image,
-										 LIGHTMAP_SIZE, LIGHTMAP_SIZE, qfalse, qfalse,
-										 (qboolean) !!r_ext_compressed_lightmaps->integer, GL_CLAMP );
+
+		// vanilla hardcoded the maximum number of lightmaps to 256
+		// ideally, we use a 2048x2048 atlas to store all of those in one texture
+		int currentAtlas = i / LIGHTMAPS_PER_ATLAS;
+		int currentAtlasLightmap = i % LIGHTMAPS_PER_ATLAS;
+		int xoff = (currentAtlasLightmap % LIGHTMAPS_PER_ATLAS_DIMENSION) * LIGHTMAP_SIZE;
```

Comment:
Lightmaps use linear filtering and no mip mapping. This (no mip mapping) is because q3map2 packs surfaces with a pixel padding of 1px to safe on image space. The second mip level would already average pixels that dont belong to the same surface which would introduce light bleeding. The first mip would already darken all uv island borders of the surfaces. Haven't looked into the q3map2 lightmap uv packing code, but I'm pretty sure the padding also applies to the image border. At least, I never encountered a map where atlasing introduced light bleeding. I'm using a similar approach with my blender addon to load lightmaps, so I've seen alot of maps from alot of different games, with no issues regarding this. Clamping is bascially never used as all the lightmap coodinates always are in range [0.0 - 1.0] already.

Reply:
Cool, forgot that lightmaps already have 1pixel padding. Should be fine then. You probably told me so on discord before but I forgot.